deps/streamsearch/sbmh.js

```diff
@@ -24,66 +24,76 @@
  * Based heavily on the Streaming Boyer-Moore-Horspool C++ implementation
  * by Hongli Lai at: https://github.com/FooBarWidget/boyer-moore-horspool
  */
-var EventEmitter = require('events').EventEmitter,
-    inherits = require('util').inherits;
+const EventEmitter = require('events').EventEmitter,
+  inherits = require('util').inherits;
 
 function SBMH(needle) {
-  if (typeof needle === 'string')
+  if (typeof needle === 'string') {
     needle = Buffer.from(needle);
-  var i, j, needle_len = needle.length;
+  }
+
+  if ((needle instanceof Buffer) === false) {
+    throw new TypeError("The needle has to be a String or a Buffer.");
+  }
+
+  const needle_len = needle.length;
+
+  if (needle_len === 0) {
+    throw new Error("The needle can not be an empty String/Buffer.");
+  }
+
+  if (needle_len > 256) {
+    throw new Error("The needle can not have a length bigger than 256.");
+  }
 
   this.maxMatches = Infinity;
   this.matches = 0;
 
-  this._occ = new Array(256);
+  this._occ = new Array(256)
+    .fill(needle_len); // Initialize occurrence table.
   this._lookbehind_size = 0;
   this._needle = needle;
   this._bufpos = 0;
 
   this._lookbehind = Buffer.alloc(needle_len);
 
-  // Initialize occurrence table.
-  for (j = 0; j < 256; ++j)
-    this._occ[j] = needle_len;
-
   // Populate occurrence table with analysis of the needle,
   // ignoring last letter.
-  if (needle_len >= 1) {
-    for (i = 0; i < needle_len - 1; ++i)
-      this._occ[needle[i]] = needle_len - 1 - i;
-  }
+  for (var i = 0; i < needle_len - 1; ++i)
```

Comment:
`i` isn't used outside the loop so we could use `let` instead of `var` here:

```suggestion
  for (let i = 0; i < needle_len - 1; ++i)
```

Reply:
Var is used for performance reasons here, because let inside a for loop creates a new instance on every iteration: https://stackoverflow.com/questions/40449494/why-is-using-let-inside-a-for-loop-so-slow-on-chrome

Allegedly this is optimized away by modern v8, but in other fastify libraries benchmarks were still showing a difference, so we typically prefer to keep var there.

Reply:
TIL, thanks for the link!